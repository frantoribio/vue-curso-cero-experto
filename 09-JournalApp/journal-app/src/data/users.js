const users = [
  {
    uid: '1',
    name: 'John Doe',
    email: 'john.doe@example.com',
    username: 'jdoe',
    password: 'password',
  },
  {
    uid: '2',
    name: 'Jane Doe',
    email: 'jane.doe@example.com',
    username: 'jdoe',
    password: 'password',
  },
]

export const getUser = () => {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve(users[0])
    }, 1000)
  })
}

export default users
