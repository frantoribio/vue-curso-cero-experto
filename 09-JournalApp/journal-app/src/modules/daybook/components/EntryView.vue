<template>
  <!-- Transicciones me pidde un solo div!!! -->
  <transition
    name="fade"
    mode="out-in"
  >
    <div v-if="myEntry">
      <div class="flex h-20 flex-row items-center justify-between p-2">
        <div>
          <span class="text-lg font-bold text-primary">{{ entryDate.day }}</span>
          <span class="mx-2 text-lg font-semibold text-base-content">{{ entryDate.month }}</span>
          <span class="mx-1 text-lg font-light text-base-content">{{ entryDate.year }}, </span>
          <span class="mx-1 text-lg font-light text-base-content">{{ entryDate.dayOfWeek }}.</span>
        </div>
        <div>
          <button class="btn btn-error mx-2 gap-2">
            <Icon
              icon="fa:trash"
              :inline="true"
              class="h-5 w-5"
            />
            Borrar
          </button>
          <button class="btn btn-primary mx-2 gap-2">
            <Icon
              icon="fa:upload"
              :inline="true"
              class="h-5 w-5"
            />
            Subir foto
          </button>
        </div>
      </div>
      <hr class="border-t border-primary" />
      <div class="flex h-full flex-col px-3">
        <textarea
          v-model.trim="myEntry.text"
          placeholder="¿Qué sucedió hoy?"
          class="ext-base-content textarea textarea-primary my-2 h-full w-full rounded-md bg-base-100"
        ></textarea>
      </div>
      <img
        :src="myEntry.picture"
        alt="imagen"
        class="fixed right-10 bottom-32 m-2 w-48 rounded-lg border-4 border-base-200 object-cover shadow-xl"
      />
      <FavButton
        icon="fa-regular:save"
        class="fixed right-10 bottom-14 h-12 w-12 shadow-md"
        @click="saveEntry()"
      />
    </div>
  </transition>
</template>

<script setup>
  import FavButton from '@/components/FabButton.vue'
  import getDateParsered from '@/helpers/dateparser.js'
  import { Icon } from '@iconify/vue'
  import { computed, ref, watch } from 'vue'
  import { useRouter } from 'vue-router'
  import JournalStore from '../stores/journal'

  const props = defineProps({
    id: {
      type: String,
      required: true,
    },
  })

  const router = useRouter()
  const journalStore = JournalStore()

  const myEntry = ref(null)

  const loadEntry = () => {
    myEntry.value = journalStore.getEntryById(props.id)
    if (!myEntry.value) {
      router.push({ name: 'daybook-no-entry' })
    }
  }

  // Y cargamos la entrada
  loadEntry()

  const saveEntry = async () => {
    console.log('saveEntry')
  }

  // La fecha es computada
  const entryDate = computed(() => {
    return getDateParsered(myEntry.value.date)
  })

  // Compruebo que existe la entrada añado un watcher!!
  watch(
    () => props.id,
    () => {
      loadEntry()
    }
  )

  // Observamos el id y sus cambios
</script>

<style lang="scss" scoped>
  .fade-enter-active,
  .fade-leave-active {
    transition: opacity 0.5s ease;
  }
  .fade-enter-from,
  .fade-leave-to {
    opacity: 0;
  }
</style>
