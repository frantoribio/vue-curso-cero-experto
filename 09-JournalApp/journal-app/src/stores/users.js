import { getUser } from '@/data/users'
import JournalStore from '@/modules/daybook/stores/journal'
import { defineStore } from 'pinia'

export const UserStore = defineStore({
  id: 'UserStore',
  state: () => ({
    user: null,
  }),

  getters: {
    isLoggedIn: (state) => state.user !== null,
  },

  actions: {
    setUser(user) {
      this.user = user
    },

    async logIn() {
      const user = await getUser()
      this.setUser(user)
      // Le cargamos las entradas del usuario!!!
      const journalStore = JournalStore()
      journalStore.fetchEntries()
    },

    async logOut() {
      await getUser() // Solo por perder el tiempo y parezca asincrono
      // Le limpiamos los datos del usuario!!!
      const journalStore = JournalStore()
      journalStore.clearData()
      this.setUser(null)
    },
  },
})

export default UserStore
