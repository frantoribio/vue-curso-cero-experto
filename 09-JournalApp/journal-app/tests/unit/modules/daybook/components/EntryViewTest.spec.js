import { beforeEach, describe, expect, test } from 'vitest'

import { createTestingPinia } from '@pinia/testing'
import { shallowMount } from '@vue/test-utils'
import { createRouterMock, injectRouterMock } from 'vue-router-mock'

import EntryView from '@/modules/daybook/components/EntryView.vue'

import entries from '../stores/mocks/entries.mocks'
import users from '../stores/mocks/users.mocks'

describe('Daybook Componente -> EntryView', () => {
  let wrapper
  const router = createRouterMock()

  beforeEach(() => {
    injectRouterMock(router)
    wrapper = shallowMount(EntryView, {
      props: {
        id: entries[0].id,
      },
      global: {
        plugins: [
          createTestingPinia({
            initialState: {
              JournalStore: { entries: entries },
              UserStore: { user: users[0] },
            },
          }),
        ],
      },
    })
  })

  test('Debe hacer match con el snapshot', () => {
    expect(wrapper.text()).toMatchSnapshot()
  })
})
