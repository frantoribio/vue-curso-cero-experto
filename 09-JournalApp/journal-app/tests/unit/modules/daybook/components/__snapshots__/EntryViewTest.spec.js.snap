// Vitest Snapshot v1

exports[`Daybook Componente -> EntryView > Debe hacer match con el snapshot 1`] = `"21Agosto2022, Domingo. Borrar  Subir foto"`;
